### PR TITLE
feat: add `job_timeout_ms` to job configuration classes

### LINKED ISSUE #1421: Expose jobTimeoutMs field in JobConfigurationQuery
At job creation time I want to set a timeout for the job if it takes too long to run. Too many long running queries eat up the slot reservation and significantly deteriorates query performance. Currently we have to manually cancel the queries in the console. I noticed that there is a `jobTimeoutMs` field in the [JobConfiguration](https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#jobconfiguration) object. This is not exposed by the Python client but is exposed by other Bigquery clients ([Golang](https://github.com/googleapis/google-cloud-go/pull/5707)). Can we add this to the Python client as well?

The alternative would be to have a timer on the client side and manually cancel the request. This feels wasteful considering the feature already exists on the server side and it's just a question of exposing it in the client.

### PR DESCRIPTION
**Pull request by chalmerlowe**
Fixes https://github.com/googleapis/python-bigquery/issues/1421

Exposes job timeout setting in Python job timeout config.
https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobConfiguration.FIELDS.job_timeout_ms

### CODE REVIEW

File: tests/system/test_query.py [hunk 1]
```diff
@@ -97,6 +97,20 @@ def test_query_w_timeout(bigquery_client, query_api_method):
...
+def test_job_timeout_ms():
+    # Confirm that default status is None.
+    job_config = bigquery.QueryJobConfig()
+    assert job_config.job_timeout_ms is None
+
+    # Confirm that integers get converted to strings.
+    job_config.job_timeout_ms = 5000
+    assert job_config.job_timeout_ms == "5000"  # int is converted to string
+
+    # Confirm that attempting to set a non-integer values will raise an Error.
```
Comment (chalmerlowe) [inline tests/system/test_query.py:109]:
```suggestion
def test_job_timeout_ms_raises_valueerror():
        # Confirm that attempting to set a non-integer values will raise an Error.
```

File: google/cloud/bigquery/job/base.py [hunk 2]
```diff
@@ -171,6 +171,33 @@ def __setattr__(self, name, value):
...
+            job_config = bigquery.QueryJobConfig( jobtimeout = 5000 )
+            or
+            job_config.jobtimeout = 5000
+
+        Raises:
+            ValueError: If ``value`` type is invalid.
+        """
+
+        # None as this is an optional parameter.
+        if self._properties.get("jobTimeoutMs"):
```
Comment (tswast) [inline google/cloud/bigquery/job/base.py:189]: Let's use the `_int_or_none` helper for this instead: https://github.com/googleapis/python-bigquery/blob/53aad826a55dd1c36d014c80ada515273e8aa92a/google/cloud/bigquery/_helpers.py#L860

BigQuery often serializes integers as strings in the JSON response.

File: google/cloud/bigquery/job/base.py [hunk 3]
```diff
@@ -171,6 +171,33 @@ def __setattr__(self, name, value):
             )
         super(_JobConfig, self).__setattr__(name, value)
 
+    @property
+    def job_timeout_ms(self):
+        """Optional parameter. Job timeout in milliseconds. If this time limit is exceeded, BigQuery might attempt to stop the job.
+        https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobConfiguration.FIELDS.job_timeout_ms
+        e.g.
+
+            job_config = bigquery.QueryJobConfig( jobtimeout = 5000 )
```
Comment (tswast) [inline google/cloud/bigquery/job/base.py:180]: Fix the property name in the code sample.

**Review by tswast — COMMENTED** [review]
A couple of suggestions, otherwise looking good. Thanks!

File: google/cloud/bigquery/job/base.py (hunk 2 above)
Comment (chalmerlowe) [reply google/cloud/bigquery/job/base.py:189]: I will see about using that. Thanks for the tip.

File: google/cloud/bigquery/job/base.py [hunk 4]
```diff
@@ -171,6 +171,33 @@ def __setattr__(self, name, value):
...
+            or
+            job_config.jobtimeout = 5000
```
Comment (chalmerlowe) [inline google/cloud/bigquery/job/base.py:182]:
```suggestion
            job_config = bigquery.QueryJobConfig( job_timeout_ms = 5000 )
            or
            job_config.job_timeout_ms = 5000
```

**Comment by dhendry** [comment]
+1 on exposing `job_timeout_ms`, this would be really helpful! Any idea on when this could get released?

File: tests/system/test_query.py [hunk 5]
```diff
@@ -97,6 +97,23 @@ def test_query_w_timeout(bigquery_client, query_api_method):
     assert bigquery_client.cancel_job(query_job) is not None
 
 
+def test_job_timeout_ms():
+    # Confirm that default status is None.
+    job_config = bigquery.QueryJobConfig()
```
Comment (chalmerlowe) [inline tests/system/test_query.py:102]:
```suggestion
    job_config = bigquery._JobConfig()
```

File: tests/system/test_query.py [hunk 6]
```diff
@@ -97,6 +97,23 @@ def test_query_w_timeout(bigquery_client, query_api_method):
     assert bigquery_client.cancel_job(query_job) is not None
 
 
+def test_job_timeout_ms():
+    # Confirm that default status is None.
+    job_config = bigquery._JobConfig()
```
Comment (chalmerlowe) [inline tests/system/test_query.py:102]:
```suggestion
    job_config = bigquery.QueryJobConfig()
```

**Comment by chalmerlowe** [comment]
> +1 on exposing `job_timeout_ms`, this would be really helpful! Any idea on when this could get released?

working on it. need to take care of ensuring that coverage testing passes.

File: tests/unit/job/test_base.py [hunk 7]
```diff
@@ -1228,3 +1228,20 @@ def test_labels_setter(self):
         job_config = self._make_one()
         job_config.labels = labels
         self.assertEqual(job_config._properties["labels"], labels)
+
+    def test_job_timeout_ms_raises_valueerror(self):
+        # Confirm that attempting to set a non-integer values will raise an Error.
+        with pytest.raises(ValueError):
+            # from google.cloud.bigquery import QueryJobConfig
```
Comment (Linchin) [inline tests/unit/job/test_base.py:1235]: nit: delete these commented out lines?

Comment (chalmerlowe) [reply tests/unit/job/test_base.py:1235]: good catch, will do.

File: tests/unit/job/test_base.py [hunk 8]
```diff
@@ -1228,3 +1228,20 @@ def test_labels_setter(self):
...
+            # job_config = QueryJobConfig()
```
Comment (chalmerlowe) [inline tests/unit/job/test_base.py:1236]:
```suggestion
            # from google.cloud.bigquery import QueryJobConfig
            # job_config = QueryJobConfig()
```

**Comment by chalmerlowe** [comment]
> Thank you @chalmerlowe for implementing this feature! I see that this is a change applied to the base class. For my education, I wonder if we need to make any changes to the subclasses, or `job_timeout_ms` is completely covered in `_JobConfig._properties`? 
To the best of my knowledge, none of the subclasses require any additions or modifications related to `job_timeout_ms`. They should inherit this property same as any other property. None of their behaviors is such that we need to change how that attribute is used by any of the subclasses.

> Also, I think it makes sense to add this to system tests too, WDYT?
Thanks for the question. I am gonna stick with unit tests on this PR.

**Review by Linchin — APPROVED** [review]

**Pull request merged**